clcache.py

```diff
@@ -376,101 +376,111 @@ def save(self):
 
 
 class CacheStatistics:
+    _defaultValue = 0
+
     def __init__(self, cacheDirectory):
         lockName = cacheDirectory.replace(':', '-').replace('\\', '-') + "_stats"
         timeoutMs = 10 * 1000
         self._lock = SystemSharedLock(lockName, timeoutMs)
+        self._cacheDirectory = cacheDirectory
 
+    def _get(self, key):
         with self._lock:
-            self._stats = PersistentJSONDict(os.path.join(cacheDirectory, "stats.txt"))
+            storage = PersistentJSONDict(os.path.join(self._cacheDirectory, "stats.txt"))
+            if not key in storage:
+                return self._defaultValue
+            else:
+                return storage[key]
 
-        for k in ["CallsWithoutSourceFile",
-                  "CallsWithMultipleSourceFiles",
-                  "CallsWithPch",
-                  "CallsForLinking",
-                  "CallsForExternalDebugInfo",
-                  "CacheEntries", "CacheSize",
-                  "CacheHits", "CacheMisses",
-                  "EvictedMisses", "HeaderChangedMisses",
-                  "SourceChangedMisses"]:
-            if k not in self._stats:
-                self._stats[k] = 0
+    def _set(self, key, value):
+        with self._lock:
+            storage = PersistentJSONDict(os.path.join(self._cacheDirectory, "stats.txt"))
+            storage[key] = value
+            storage.save()
+
+    def _increase(self, key, ammount):
+        with self._lock:
+            storage = PersistentJSONDict(os.path.join(self._cacheDirectory, "stats.txt"))
+            if not key in storage:
+                storage[key] = self._defaultValue
+            storage[key] += ammount
+            storage.save()
```

Comment:
`_set` and `_increase` are notably similiar. I think you could fix the code duplication here by introducing a generic `_modify` method which is parametrised on the code using which the key in the dictionary is modified:

    def _modify(self, key, fn):
        with self._lock:
            storage = PersistentJSONDict(os.path.join(self._cacheDirectory, "stats.txt"))
            storage[key] = fn(storage.key(key) or self._defaultValue)
            storage.save()

(Writing this code makes me think that it should be possible to use `PersistentJSONDict` with `with`...).

With this methods at hand, `_set` and `_increase` are just:

    def _set(self, key, value):
        self._modify(key, lambda _: value)

    def _increase(self, key, amount):
        self._modify(key, lambda x: x + amount)